crates/livesplit-hotkey/src/wasm_web/mod.rs

```diff
@@ -52,10 +55,84 @@ impl Hook {
         }) as Box<dyn FnMut(KeyboardEvent)>);
 
         window
-            .add_event_listener_with_callback("keypress", callback.as_ref().unchecked_ref())
+            .add_event_listener_with_callback(
+                "keypress",
+                keyboard_callback.as_ref().unchecked_ref(),
+            )
+            .map_err(|_| Error::FailedToCreateHook)?;
+
+        let hotkey_map = hotkeys.clone();
+
+        const TOTAL_BUTTONS: usize = 20;
+        static GAMEPAD_BUTTONS: [KeyCode; TOTAL_BUTTONS] = [
+            KeyCode::Gamepad0,
+            KeyCode::Gamepad1,
+            KeyCode::Gamepad2,
+            KeyCode::Gamepad3,
+            KeyCode::Gamepad4,
+            KeyCode::Gamepad5,
+            KeyCode::Gamepad6,
+            KeyCode::Gamepad7,
+            KeyCode::Gamepad8,
+            KeyCode::Gamepad9,
+            KeyCode::Gamepad10,
+            KeyCode::Gamepad11,
+            KeyCode::Gamepad12,
+            KeyCode::Gamepad13,
+            KeyCode::Gamepad14,
+            KeyCode::Gamepad15,
+            KeyCode::Gamepad16,
+            KeyCode::Gamepad17,
+            KeyCode::Gamepad18,
+            KeyCode::Gamepad19,
+        ];
+        let mut states = Vec::new();
+        let navigator = window.navigator();
+
+        let gamepad_callback = Closure::wrap(Box::new(move || {
+            if let Ok(gamepads) = navigator.get_gamepads() {
+                let gamepads_len = gamepads.length() as usize;
+                if states.len() < gamepads_len {
+                    states.resize(gamepads_len, [false; TOTAL_BUTTONS]);
+                }
+                for (gamepad, states) in gamepads.iter().zip(&mut states) {
+                    if let Ok(gamepad) = gamepad.dyn_into::<Gamepad>() {
+                        for ((button, code), state) in gamepad
+                            .buttons()
+                            .iter()
+                            .zip(GAMEPAD_BUTTONS.iter().copied())
+                            .zip(states.iter_mut())
+                        {
+                            if let Ok(button) = button.dyn_into::<GamepadButton>() {
+                                let pressed = button.pressed();
+                                if pressed && !*state {
+                                    if let Some(callback) =
+                                        hotkey_map.lock().unwrap().get_mut(&code)
+                                    {
+                                        callback();
+                                    }
+                                }
+                                *state = pressed;
+                            }
+                        }
+                    }
+                }
+            }
+        }) as Box<dyn FnMut()>);
+
+        let interval_id = window
+            .set_interval_with_callback_and_timeout_and_arguments_0(
+                gamepad_callback.as_ref().unchecked_ref(),
+                1000 / 60,
```

Comment:
Should the refresh rate (60 fps) be extracted as a constant?

Reply:
Considering it's not used anywhere else, I don't think so. Constants for the sake of constants is not useful imo.